posts/change-data-capture-cloudsql-postgres-with-apache-airflow-1.md

```diff
@@ -0,0 +1,1002 @@
+```
+
+Based on this example, we can run:
+
+```bash
+astro dev run render change_data_capture_cloud_sql export_table_1 '2021-01-14 13:00:00'
+```
+
+Afterwards, we should see our templating print out:
+
+```bash
+# ----------------------------------------------------------
+# property: export_sql
+# ----------------------------------------------------------
+SELECT * FROM some_schema.table_1
+WHERE event_timestamp AT TIME ZONE 'UTC' >= '2021-01-14T12:00:00+00:00' AT TIME ZONE 'UTC'
+AND event_timestamp AT TIME ZONE 'UTC' < '2021-01-14T13:00:00+00:00'
+
+# ----------------------------------------------------------
+# property: gcs_bucket
+# ----------------------------------------------------------
+your_gcs_bucket
+
+# ----------------------------------------------------------
+# property: gcs_keypath
+# ----------------------------------------------------------
+your_gcs_bucket/cloudsql_export/raw/table_1/2021/1/14/table_1_20210114T130000.csv
+```
+
+We see our SQL Query, our bucket, and the full `gcs_keypath` fully rendered based on our `execution_date` chosen. Try the same for `get_schema`. When complete, the success of such signal this DAG is now ready to be run. 
+
+## Step 12: Implement Scheduling and Timing
+
+Scheduling and timing in Airflow possesses a few relative time-based parameters that need to be defined and thought through together to fully grasp. In `Step 9`, we introduced one of them: the `execution_date`. Now we must add in the interval those are scheduled to execute in, called the `schedule_interval`. It's important to understand how they relate together with the time it is on the clock.
+
+When a DAG starts at its scheduled time, it begins at the end of the interval chosen. So for `schedule_interval='@hourly'` and an `execution_date` of `2021-01-14T13:00:00+00:00`, the DAG would actually be starting at `2021-01-14T14:00:00+00:00`. 
+
+For a visual example, the image below shows the task tree of a particular DAG in which this concept can be observed. The first line, `Run: 2021-01-19, 21:00:00 UTC`, shows the `execution_date` from an hourly scheduled DAG, set as `schedule_interval='@hourly'`. Notice `Started: 2021-01-19T22:00:16xxx` is an hour (a single interval) after the schedule date. That is, it starts one `schedule_interval` later than the `execution_date`.
+
+<span style="display:block;text-align:center;">
+  <img src="../assets/cdc-cloudsql-1/cdc_cloudsql_airflow_2021-01-19_at_2.48.37_PM.png" width=400> 
+</span>
+
+The `start_date` you set in the DAG definition follows this logic. If you choose:
+
+```python
+start_date = DateTime(2021, 1, 14, 13, 0, 0),
+schedule_interval = '@hourly',
+catchup = False
+```
+
+you would see the first run at `2021-01-14T14:00:00+00:00` if the current date were before that and before the `start_date`. If the current calendar date was after the `start_date`, you would see the first run start at the beginning of the most recent hour. 
```

Comment:
Capitalize the You.

Reply:
Actually, would change this sentence to be parallel to the following one.

"If the current calendar date were before...., you would see the first run at xxx. Conversely, if the current calendar date were after..."

Reply:
Agreed it's tough. There's 3 dates happening though: The current calendar date, the start_date, and the execution_date.